Use errors.As to detect wrapping in StatusCause

cbandy opened the pull request:
#### What type of PR is this?

/kind bug

#### What this PR does / why we need it:

Two functions in this package, `StatusCause` and `HasStatusCause`, do not handle wrapped errors. With this, the entire package handles wrapped errors consistently.

#### Which issue(s) this PR fixes:

Fixes #108528

#### Special notes for your reviewer:

#### Does this PR introduce a user-facing change?

Users of these functions will have already done their own unwrapping and are unaffected.

```release-note
NONE
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

<!--
This section can be blank if this pull request does not require a release note.

When adding links which point to resources within git repositories, like
KEPs or supporting documentation, please reference a specific commit and avoid
linking directly to the master branch. This ensures that links reference a
specific point in time, rather than a document that may change over time.

See here for guidance on getting permanent links to files: https://help.github.com/en/articles/getting-permanent-links-to-files

Please use the following format for linking documentation:
- [KEP]: <link>
- [Usage]: <link>
- [Other doc]: <link>
-->

cbandy committed:
Use errors.As to detect wrapping in StatusCause

Other functions in this package handle wrapped errors since 3244350046.

tallclair commented on staging/src/k8s.io/apimachinery/pkg/api/errors/errors_test.go:
```diff
@@ -625,3 +625,31 @@ func TestIsErrorTypesByReasonAndCode(t *testing.T) {
+	nested := fmt.Errorf("twice: %w", wrapped)
+	if cause, ok := StatusCause(nested, "SomeCause"); !ok || cause != err.ErrStatus.Details.Causes[0] {
+		t.Errorf("expected cause when nested, got %v: %#v", ok, cause)
+	}
```
Add a test case where you pass a non-status error

cbandy replied:
Added in amended commit.

tallclair reviewed (commented):
missing a test case, otherwise lgtm

cbandy commented:
Thanks for the quick review! I've addressed all comments.

tallclair commented on staging/src/k8s.io/apimachinery/pkg/api/errors/errors.go:
```diff
@@ -87,21 +87,21 @@ func (e *StatusError) DebugError() (string, []interface{}) {
-		if cause.Type == name {
-			return cause, true
+	var status APIStatus
+	if errors.As(err, &status) && status.Status().Details != nil {
```
This is probably splitting hairs, but I was surprised to see that `errors.As` is implemented using reflection (notoriously slow). A quick benchmark confirmed that it's about an order of magnitude slower:
```
goos: linux
goarch: amd64
pkg: k8s.io/apimachinery/pkg/api/errors
cpu: 11th Gen Intel(R) Core(TM) i7-1185G7 @ 3.00GHz
BenchmarkTypeAssert_pos-8   	300704044	         3.950 ns/op
BenchmarkTypeAssert_neg-8   	277394091	         4.543 ns/op
BenchmarkErrorsAs_pos-8     	10509163	       118.8 ns/op
BenchmarkErrorsAs_neg-8     	35966838	        31.30 ns/op
```

I'm not sure if this is called in any hot paths, but I think it would be safest to just do the type assertion first, and then fallback to `errors.As`.

I'm honestly not sure how nitpicky this is, so feel free to weight for an apimachinery maintainer to weigh in.

cbandy replied:
Understood. `errors.As` was introduced to other functions in this package during #88629. I don't see any discussion there about performance or why `StatusCause` might have been excluded.

pacoxu commented:
/test pull-kubernetes-unit

tallclair commented:
/lgtm

leilajal commented:
/triage accepted

cbandy committed:
api/errors: Improve performance of Is* functions

Type assertion is faster than "errors.As" when the error is not wrapped.
Allocations are unaffected.

Benchmark results:
```
name                                    old time/op    new time/op    delta
IsAlreadyExistsWrappedErrors/Nil-4        72.8ns ±11%    79.6ns ±12%     ~     (p=0.343 n=4+4)
IsAlreadyExistsWrappedErrors/Bare-4        336ns ± 6%      95ns ±10%  -71.70%  (p=0.029 n=4+4)
IsAlreadyExistsWrappedErrors/Wrapped-4     416ns ± 1%     426ns ± 4%     ~     (p=0.486 n=4+4)
IsNotFoundWrappedErrors/Nil-4              102ns ± 4%      97ns ± 4%     ~     (p=0.114 n=4+4)
IsNotFoundWrappedErrors/Bare-4             350ns ± 3%     102ns ± 6%  -70.78%  (p=0.029 n=4+4)
IsNotFoundWrappedErrors/Wrapped-4          448ns ± 3%     462ns ± 2%     ~     (p=0.200 n=4+4)
```

cbandy commented:
I've rebased and added a commit that uses type assertion to short-circuit calls to `errors.As`. Does this address your concern @tallclair?

```
name                                    old time/op    new time/op    delta
IsAlreadyExistsWrappedErrors/Nil-4        72.8ns ±11%    79.6ns ±12%     ~     (p=0.343 n=4+4)
IsAlreadyExistsWrappedErrors/Bare-4        336ns ± 6%      95ns ±10%  -71.70%  (p=0.029 n=4+4)
IsAlreadyExistsWrappedErrors/Wrapped-4     416ns ± 1%     426ns ± 4%     ~     (p=0.486 n=4+4)
IsNotFoundWrappedErrors/Nil-4              102ns ± 4%      97ns ± 4%     ~     (p=0.114 n=4+4)
IsNotFoundWrappedErrors/Bare-4             350ns ± 3%     102ns ± 6%  -70.78%  (p=0.029 n=4+4)
IsNotFoundWrappedErrors/Wrapped-4          448ns ± 3%     462ns ± 2%     ~     (p=0.200 n=4+4)

name                                    old alloc/op   new alloc/op   delta
IsAlreadyExistsWrappedErrors/Nil-4         16.0B ± 0%     16.0B ± 0%     ~     (all equal)
IsAlreadyExistsWrappedErrors/Bare-4        16.0B ± 0%     16.0B ± 0%     ~     (all equal)
IsAlreadyExistsWrappedErrors/Wrapped-4     16.0B ± 0%     16.0B ± 0%     ~     (all equal)
IsNotFoundWrappedErrors/Nil-4              16.0B ± 0%     16.0B ± 0%     ~     (all equal)
IsNotFoundWrappedErrors/Bare-4             16.0B ± 0%     16.0B ± 0%     ~     (all equal)
IsNotFoundWrappedErrors/Wrapped-4          16.0B ± 0%     16.0B ± 0%     ~     (all equal)

name                                    old allocs/op  new allocs/op  delta
IsAlreadyExistsWrappedErrors/Nil-4          1.00 ± 0%      1.00 ± 0%     ~     (all equal)
IsAlreadyExistsWrappedErrors/Bare-4         1.00 ± 0%      1.00 ± 0%     ~     (all equal)
IsAlreadyExistsWrappedErrors/Wrapped-4      1.00 ± 0%      1.00 ± 0%     ~     (all equal)
IsNotFoundWrappedErrors/Nil-4               1.00 ± 0%      1.00 ± 0%     ~     (all equal)
IsNotFoundWrappedErrors/Bare-4              1.00 ± 0%      1.00 ± 0%     ~     (all equal)
IsNotFoundWrappedErrors/Wrapped-4           1.00 ± 0%      1.00 ± 0%     ~     (all equal)
```

cbandy commented:
/assign @lavalamp

This has been reviewed once, but I rebased with another change.

tallclair commented:
/ok-to-test

lavalamp commented:
Is this intended for clients or for within apiserver code? Because if apiserver wraps one of these before returning it to clients, that's a bug that we need to fix.

cbandy commented:
My use case is strictly on the client-side. My code makes API calls using `k8s.io/go-client` and wraps errors as they return up the call stack.

#108528, specifically, my high-level error handler was expecting `HasStatusCause(err, corev1.NamespaceTerminatingCause)` to unwrap in a manner similar to `Is*` functions.

cbandy commented:
@lavalamp Is there anything else I should do on this one?

lavalamp commented on staging/src/k8s.io/apimachinery/pkg/api/errors/errors.go:
```diff
@@ -757,7 +757,8 @@ func IsRequestEntityTooLargeError(err error) bool {
 // It supports wrapped errors and returns false when the error is nil.
 func IsUnexpectedServerError(err error) bool {
-	if status := APIStatus(nil); errors.As(err, &status) && status.Status().Details != nil {
+	status, ok := err.(APIStatus)
```
Are these functions tested? It seems that this changes the type of `status` and I suspect they don't do exactly the same thing anymore?

lavalamp replied:
I guess APIStatus is an interface type, let me rethink this

lavalamp commented:
OK I think this is safe, thanks.

/lgtm
/approve
